source/extensions/common/matcher/trie_matcher.h

```diff
@@ -0,0 +1,151 @@
+#pragma once
+
+#include "envoy/config/common/matcher/v3/matcher.pb.h"
+#include "envoy/config/common/matcher/v3/matcher.pb.validate.h"
+#include "envoy/matcher/matcher.h"
+#include "envoy/network/filter.h"
+#include "envoy/server/factory_context.h"
+
+#include "source/common/network/lc_trie.h"
+#include "source/common/network/utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Matcher {
+
+using ::Envoy::Matcher::DataInputFactoryCb;
+using ::Envoy::Matcher::DataInputGetResult;
+using ::Envoy::Matcher::DataInputPtr;
+using ::Envoy::Matcher::MatchState;
+using ::Envoy::Matcher::MatchTree;
+using ::Envoy::Matcher::OnMatch;
+using ::Envoy::Matcher::OnMatchFactory;
+using ::Envoy::Matcher::OnMatchFactoryCb;
+
+template <class DataType> struct TrieNode {
+  size_t index_;
+  uint32_t prefix_len_;
+  bool exclusive_;
+  std::shared_ptr<OnMatch<DataType>> on_match_;
+
+  friend bool operator==(const TrieNode<DataType>& lhs, const TrieNode<DataType>& rhs) {
+    return lhs.index_ == rhs.index_ && lhs.prefix_len_ == rhs.prefix_len_ &&
+           lhs.exclusive_ == rhs.exclusive_ && lhs.on_match_ == rhs.on_match_;
+  }
+  template <typename H>
+  friend H AbslHashValue(H h, // NOLINT(readability-identifier-naming)
+                         const TrieNode<DataType>& node) {
+    return H::combine(std::move(h), node.index_, node.prefix_len_, node.exclusive_, node.on_match_);
+  }
+};
+
+template <class DataType> struct TrieNodeComparator {
+  inline bool operator()(const TrieNode<DataType>& lhs, const TrieNode<DataType>& rhs) const {
+    if (lhs.prefix_len_ > rhs.prefix_len_) {
+      return true;
+    }
+    if (lhs.prefix_len_ == rhs.prefix_len_ && lhs.index_ < rhs.index_) {
+      return true;
+    }
+    return false;
+  }
+};
+
+/**
+ * Implementation of a `sublinear` LC-trie matcher.
+ */
+template <class DataType> class TrieMatcher : public MatchTree<DataType> {
+public:
+  TrieMatcher(DataInputPtr<DataType>&& data_input,
+              std::unique_ptr<Network::LcTrie::LcTrie<TrieNode<DataType>>>&& trie)
+      : data_input_(std::move(data_input)), trie_(std::move(trie)) {}
+
+  typename MatchTree<DataType>::MatchResult match(const DataType& data) override {
+    const auto input = data_input_->get(data);
+    if (input.data_availability_ == DataInputGetResult::DataAvailability::NotAvailable) {
+      return {MatchState::UnableToMatch, absl::nullopt};
+    }
+    if (!input.data_) {
+      return {MatchState::MatchComplete, absl::nullopt};
+    }
+    const Network::Address::InstanceConstSharedPtr addr =
+        Network::Utility::parseInternetAddressNoThrow(*input.data_);
+    if (!addr) {
+      return {MatchState::MatchComplete, absl::nullopt};
+    }
+    auto values = trie_->getData(addr);
+    std::sort(values.begin(), values.end(), TrieNodeComparator<DataType>());
```

Comment:
Comment here would be helpful, I'm guessing this ends up sorting by most specific match or something?

Reply:
Done.